src/pyEQL/solution.py

```diff
@@ -37,6 +37,87 @@
 UNKNOWN_OXI_STATE = "unk"
 K_W = 1e-14  # ion product of water at 25 degC
 
+from pyEQL import ureg
+
+# Extend the UnitRegistry with custom units
+ureg.define('ppb = 1e-9 * mole / mole = parts_per_billion')
+ureg.define('ppm = 1e-6 * mole / mole = parts_per_million')
+ureg.define('% = 0.01 * mole / mole = percent')
+
+
+def preprocess_units(amount: str) -> str:
+    """
+    Preprocess the unit string to handle custom units like 'ppb' and '%'.
+    Args:
+        amount (str): The input amount with units (e.g., '1000 ppb', '0.1%').
+    Returns:
+        str: A modified amount string compatible with Pint.
+    """
+    if "ppb" in amount:
+        return amount.replace("ppb", "microgram/liter")
+    elif "%" in amount:
+        # Convert '%' to a dimensionless fraction
+        value, unit = amount.split()
+        return f"{float(value) / 100} dimensionless"
+    return amount  # Return unchanged if no custom units
+
+def add_solute(self, formula: str, amount: str):
+    """
+    Primary method for adding substances to a pyEQL solution.
+    """
+    # Preprocess the amount to handle custom units
+    amount = preprocess_units(amount)
+
+    if ureg.Quantity(amount).dimensionality in (
+        "[substance]/[length]**3",
+        "[mass]/[length]**3",
+    ):
+        orig_volume = self.volume
+        quantity = ureg.Quantity(amount)
+        mw = self.get_property(formula, "molecular_weight")
+        target_mol = quantity.to("moles", "chem", mw=mw, volume=self.volume, solvent_mass=self.solvent_mass)
+        self.components[formula] = target_mol.to("moles").magnitude
+        solute_vol = self._get_solute_volume()
+        target_vol = orig_volume - solute_vol
+        target_mass = target_vol * ureg.Quantity(self.water_substance.rho, "g/L")
+        mw = self.get_property(self.solvent, "molecular_weight")
+        if mw is None:
+            raise ValueError(f"Molecular weight for solvent {self.solvent} not found in database. Cannot proceed.")
+        target_mol = target_mass.to("g") / mw.to("g/mol")
+        self.components[self.solvent] = target_mol.magnitude
+    else:
+        quantity = ureg.Quantity(amount)
+        mw = ureg.Quantity(self.get_property(formula, "molecular_weight"))
+        target_mol = quantity.to("moles", "chem", mw=mw, volume=self.volume, solvent_mass=self.solvent_mass)
+        self.components[formula] = target_mol.to("moles").magnitude
+        self.volume_update_required = True
```

Comment:
I'm not sure why there is a copy of `add_solute` added at the top of the file; this should be removed and we should just use the one attached to the `Solution` class, below.